components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -0,0 +1,336 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/asset_discovery_manager.h"
+
+#include <utility>
+
+#include "base/strings/strcat.h"
+#include "brave/components/brave_wallet/browser/blockchain_registry.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_service.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/json_rpc_service.h"
+#include "brave/components/brave_wallet/browser/keyring_service.h"
+#include "brave/components/brave_wallet/browser/pref_names.h"
+#include "brave/components/brave_wallet/common/eth_address.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+AssetDiscoveryManager::AssetDiscoveryManager(BraveWalletService* wallet_service,
+                                             JsonRpcService* json_rpc_service,
+                                             KeyringService* keyring_service,
+                                             PrefService* prefs)
+    : wallet_service_(wallet_service),
+      json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs),
+      weak_ptr_factory_(this) {
+  keyring_service_->AddObserver(
+      keyring_service_observer_receiver_.BindNewPipeAndPassRemote());
+}
+
+AssetDiscoveryManager::~AssetDiscoveryManager() = default;
+
+void AssetDiscoveryManager::DiscoverAssets(
+    const std::string& chain_id,
+    mojom::CoinType coin,
+    const std::vector<std::string>& account_addresses,
+    bool triggered_by_accounts_added,
+    const std::string& from_block,
+    const std::string& to_block,
+    const std::vector<std::string>& supported_chain_ids,
+    DiscoverAssetsCompletedCallback callback) {
+  // Asset discovery only supported on select EVM chains
+  if (coin != mojom::CoinType::ETH ||
+      !base::Contains(supported_chain_ids, chain_id)) {
+    CompleteDiscoverAssets(
+        chain_id, std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kMethodNotSupported,
+        l10n_util::GetStringUTF8(IDS_WALLET_METHOD_NOT_SUPPORTED_ERROR),
+        triggered_by_accounts_added);
+    return;
+  }
+
+  // Asset discovery only supported when using Infura proxy
```

Comment:
Can you create a follow up issue that always issue these requests to our Infura proxy even when users change the RPC URL for preloading networks?

Reply:
https://github.com/brave/brave-browser/issues/26859